Rebase develop onto 0.21.1

@ikalash, I think it's a good time to rebase onto the latest spack release (0.21.1).

I have a branch here that does that:
https://github.com/xylar/spack/tree/rebase-onto-0.21.1

The rebase only had conflicts for trilinos for one commit (explicitly adding cmake as a dependency to add a lower bound) and fixing that seemed quite straightforward.  Also, we don't actually use that the `trilinos` package, we use `trilinos-for-albany` so I don't think there's any risk at all.

Let me know when the timing is right for me to force-push to `develop`.  Obviously, let me know if you see any problems either in advance or during testing.

It looks like there are issues with PyAlbany building now in the spack build, after @xylar's merge/force push.  The errors look as follows: 

```
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

80 errors found in build log:
     1675    In file included from /tmp/ikalash/spack-stage/spack-stage-albany-
             develop-arkng5gny6zpzlqkykyarnjtdjjau3ml/spack-build-arkng5g/tpls/
             omega_h/pybind11-src/include/pybind11/attr.h:13,
     1676                     from /tmp/ikalash/spack-stage/spack-stage-albany-
             develop-arkng5gny6zpzlqkykyarnjtdjjau3ml/spack-build-arkng5g/tpls/
             omega_h/pybind11-src/include/pybind11/pybind11.h:45,
     1677                     from /tmp/ikalash/spack-stage/spack-stage-albany-
             develop-arkng5gny6zpzlqkykyarnjtdjjau3ml/spack-src/pyAlbany/src/Al
             bany_Pybind11_Include.hpp:1,
     1678                     from /tmp/ikalash/spack-stage/spack-stage-albany-
             develop-arkng5gny6zpzlqkykyarnjtdjjau3ml/spack-src/pyAlbany/src/Al
             bany_Pybind11_Comm.hpp:14,
     1679                     from /tmp/ikalash/spack-stage/spack-stage-albany-
             develop-arkng5gny6zpzlqkykyarnjtdjjau3ml/spack-src/pyAlbany/src/Al
             bany_Pybind11_Comm.cpp:7:
     1680    /tmp/ikalash/spack-stage/spack-stage-albany-develop-arkng5gny6zpzl
             qkykyarnjtdjjau3ml/spack-build-arkng5g/tpls/omega_h/pybind11-src/i
             nclude/pybind11/cast.h: In function 'std::string pybind11::detail:
             :error_string()':
  >> 1681    /tmp/ikalash/spack-stage/spack-stage-albany-develop-arkng5gny6zpzl
             qkykyarnjtdjjau3ml/spack-build-arkng5g/tpls/omega_h/pybind11-src/i
             nclude/pybind11/cast.h:446:36: error: invalid use of incomplete ty
             pe 'PyFrameObject' {aka 'struct _frame'}
     1682      446 |                 "  " + handle(frame->f_code->co_filename).
             cast<std::string>() +
     1683          |                                    ^~
```
@kliegeois do you know what would cause this?  It isn't happening in the regular Albany builds.

--- a/var/spack/repos/builtin/packages/albany/package.py
+++ b/var/spack/repos/builtin/packages/albany/package.py
@@ -80,7 +80,7 @@ class Albany(CMakePackage):
     depends_on("trilinos-for-albany@develop", when="@develop")
     depends_on("trilinos-for-albany@compass-2023-08-03", when="@compass-2023-08-03")
 
-    extends("python",             when="+py")
+    extends("python@:3.9",        when="+py")
 
     depends_on("py-pybind11",     when="+py")
     depends_on("py-numpy",        when="+py")
@@ -125,6 +125,8 @@ class Albany(CMakePackage):
                            "ON" if "+mpas" in spec else "OFF"),
                        "-DENABLE_ALBANY_PYTHON:BOOL=%s" % (
                            "ON" if "+py" in spec else "OFF"),
+                       "-DPYTHON_EXECUTABLE=%s" % (
+                           spec["python"].command.path if "+py" in spec else ""),
                        "-DENABLE_ALBANY_EPETRA:BOOL=%s" % (
                            "ON" if "+epetra" in spec else "OFF"),
                        "-DENABLE_MESH_DEPENDS_ON_PARAMETERS:BOOL=%s" % (

